### ISSUE TITLE
Question about the changes of accuracy (mAP) on my own dataset

### ISSUE THREAD

**Issue by zhengshangdong** [facebookresearch/detr#230]
Dear sir!
          Thank you for your great job.
          I have a question with training on my own dataset. My dataset have enough data (at least 100K) and the number of classes is 30 (classes id from 0 to 29).  So, the num_classes = 30 in detr.py. I set +1 (here correspond to the no-object class (you said DETR reserves the last class id for it)) in self.class_embed = nn.Linear(hidden_dim, num_classes + 1) . I'm sure that the parameters are right.
        
         Learning_rate = 0.0001, num_queries = 100 and the other parameters are not changed. All the losses seem normal except class_error (goes from 0.00 to 100.00, e.g. 15.00, 33.33, 50.00, 66.67, 75.00) and 100 is common. Maybe there are no more than four objects in one image.

        Evaluation is performed for each epoch and the results like that (for AP@IoU=0.50)
        epoch | mAP
               -- | --
               1 | 0.018
               2 | 0.095
               3 | 0.117
               4 | 0.196
               5 | 0.218
             10 | 0.305
             20 | 0.567
             21 | 0.598 
             22 | 0.617 
             23 | 0.667 (the highest mAP) However the class error is also abnormal (from 0.00 to 100.00)
             24 | 0.574 
             25 | 0.507 
             26 | 0.486 
             27 | 0.504 
             26 | 0.431 

         I didn't stop the training process and the existing results are shown above. 
         I'm not sure this training situation is normal or not. So I don't know if it makes sense to go on training.
         Is there a decrease in accuracy (mAP) when you trained the COCO dataset? Do you think I should adjust any parameters (e.g. learning_rate, learning_rate_decay). 

        Thank you for your time and best wishes to you!

**Comment by zhengshangdong** [comment]
Additionally, in detr.py your code: num_classes = 20 if args.dataset_file != 'coco' else 91. 

That's say the num_classes = 91 when we trained the coco dataset (you said the num_classes must set to maximal class label + 1 and the maximal class label in coco dataset is 90). I understand the num_classes=90 + 1 is considering the no-object.

However, in detr.py your code: self.class_embed = nn.Linear(hidden_dim, num_classes + 1). 
I don't know why should we +1 in this code. The num_classes is equal to 92 in this way. Is that redundant?

Thank for your time and best wishes to you!

**Comment by lessw2020** [comment]
Hi @zhengshangdong,
In my experience, you should just use your max class id as num_classes.
Detr will use the next integer (+1) for its internal no obj class.

Regarding your results getting worse during training. You should drop the learning rate at that point or around that epoch range, and you should see improvements after that.
Hope that helps,
Less

**Comment by zhengshangdong** [comment]
> Hi @zhengshangdong,
> In my experience, you should just use your max class id as num_classes.
> Detr will use the next integer (+1) for its internal no obj class.
> 
> Regarding your results getting worse during training. You should drop the learning rate at that point or around that epoch range, and you should see improvements after that.
> Hope that helps,
> Less

Thank you so much for your reply!
I will try to drop the learning rate around that epoch range to expect a good results. However, I have another question about the num_classes.

As you say, the num_classes in my code should set to num_classes=29 rather than num_classes=30 (just use my max class id as num_classes. In my dataset, the class id is from 0 to 29). It is a little confusing. Doesn't num_classes represent the number of classes in the dataset?

**Comment by alcinos** [comment]
Hi @zhengshangdong 
For questions about the num_classes, please refer to #108 and comment there if something is still unclear after you read the information provided.
For advise on how to improve your training, refer to #9 

I'm closing this as it is a duplicate of either of the two issues mentioned above. Feel free to reach out if you have further concerns.

**Comment by Hrqingqing** [comment]
> Dear sir!
> Thank you for your great job.
> I have a question with training on my own dataset. My dataset have enough data (at least 100K) and the number of classes is 30 (classes id from 0 to 29). So, the num_classes = 30 in detr.py. I set +1 (here correspond to the no-object class (you said DETR reserves the last class id for it)) in self.class_embed = nn.Linear(hidden_dim, num_classes + 1) . I'm sure that the parameters are right.
> 
> ```
>      Learning_rate = 0.0001, num_queries = 100 and the other parameters are not changed. All the losses seem normal except class_error (goes from 0.00 to 100.00, e.g. 15.00, 33.33, 50.00, 66.67, 75.00) and 100 is common. Maybe there are no more than four objects in one image.
> 
>     Evaluation is performed for each epoch and the results like that (for AP@IoU=0.50)
>     epoch | mAP
>            -- | --
>            1 | 0.018
>            2 | 0.095
>            3 | 0.117
>            4 | 0.196
>            5 | 0.218
>          10 | 0.305
>          20 | 0.567
>          21 | 0.598 
>          22 | 0.617 
>          23 | 0.667 (the highest mAP) However the class error is also abnormal (from 0.00 to 100.00)
>          24 | 0.574 
>          25 | 0.507 
>          26 | 0.486 
>          27 | 0.504 
>          26 | 0.431 
> 
>      I didn't stop the training process and the existing results are shown above. 
>      I'm not sure this training situation is normal or not. So I don't know if it makes sense to go on training.
>      Is there a decrease in accuracy (mAP) when you trained the COCO dataset? Do you think I should adjust any parameters (e.g. learning_rate, learning_rate_decay). 
> 
>     Thank you for your time and best wishes to you!
> ```

Hello,I have the same problem(class_error is always 100).And you said "Maybe there are no more than four objects in one image".I think I can't understand it,could you tell me why is the number four?Thank you!
